modules/internal/ChapelArray.chpl

```diff
@@ -3988,49 +4016,119 @@ module ChapelArray {
+    // at hand...
+    //
+    // The resulting array grows dynamically. That wouldn't always be
+    // necessary if we had better shape information in iterator records.
+    // Additionally, the dynamic growing limits parallelism here, so
+    // adding better shape information would allow it to be parallel and
+    // thus enable better performance.
 
-        yield ee;
-      }
-    }
+    var i  = 0;
+    var size:size_t = 0;
+    var data:_ddata(iteratorToArrayElementType(ir.type)) = nil;
 
-    pragma "no copy"
-    var irc  = _ir_copy_recursive(ir);
+    var callAgain: bool;
+    var subloc = c_sublocid_none;
+
+    for elt in ir {
+
+      // Future: we should generally remove this copy.
+      // Note though that in some cases it invokes this function
+      // recursively - in that case it shouldn't be removed!
+      pragma "no auto destroy"
+      pragma "no copy"
+      var eltCopy = chpl__initCopy(elt);
+
+      pragma "insert auto destroy"
+      type eltType = eltCopy.type;
+
+      if i >= size {
+        // Allocate a new buffer and then copy.
+        var oldSize = size;
+        var oldData = data;
 
-    var i    = 1;
-    var size = 4;
+        if size == 0 then
+          size = 4;
+        else
+          size = 2 * size;
 
-    pragma "insert auto destroy"
-    var D    = {1..size};
+        // data allocation should match DefaultRectangular
+        __primitive("array_alloc", data, size, subloc,
+                    c_ptrTo(callAgain), c_nil);
 
-    // note that _getIterator is called in order to copy the iterator
-    // class since for arrays we need to iterate once to get the
-    // element type (at least for now); this also means that if this
-    // iterator has side effects, we will see them; a better way to
-    // handle this may be to get the static type (not initialize the
-    // array) and use a primitive to set the array's element; that may
-    // also handle skyline arrays
-    var A: [D] iteratorIndexType(irc);
+        // Now copy the data over
+        for i in 0..#oldSize {
+          // this is a move, transfering ownership
+          __primitive("=", data[i], oldData[i]);
+        }
 
-    for e in irc {
-      // The resulting array grows dynamically
-      if i > size {
-        size = 2 * size;
-        D    = { 1 .. size };
+        // Then, free the old data
+        _ddata_free(oldData, oldSize);
       }
 
-      A(i) = e;
-      i    = i + 1;
+      // Now move the element to the array
+      // The intent here is to transfer ownership to the array.
+      __primitive("=", data[i], eltCopy);
+
+      i += 1;
     }
 
-    D = { 1 .. i - 1 };
+    // let the comm layer adjust array allocation
+    if callAgain then
+      __primitive("array_alloc", data, size, subloc, c_nil, data);
 
-    return A;
+
+    if data != nil {
+      // Now construct a DefaultRectangular array using the data
+      pragma "insert auto destroy"
+      var D = { 1 .. #i };
```

Comment:
I am curious: why `pragma "insert auto destroy"` on `D` ? The user can write this code without such a pragma, right?

    proc foo() {
      var D = {1..2};
      var A: [D] int;
      return A;
    }

Reply:
Yes. But we've historically had special treatment for 'initCopy' functions.
I don't know that this pragma is required, but such a pragma was in the previous version, and I'm getting myself into enough trouble as it is, so I'd like to leave it alone for now.